variable/variable_factory.cpp

```diff
@@ -6,6 +6,30 @@
 
 namespace scipp::variable {
 
+DType VariableFactory::bin_dtype(
+    const std::vector<VariableConstView> &vars) const noexcept {
+  // First pass: Find, e.g., bucket<DataArray>, since if we have such an input
+  // we need to keep metadata in the output. Note that we would also like to
+  // prioritize bucket<Dataset> higher than bucket<DataArray>, but the former
+  // is not supported in binary operations, so this is not relevant right now.
+  if (const auto it = std::find_if(vars.begin(), vars.end(),
+                                   [](const auto &v) {
+                                     return variable::is_bins(v) &&
+                                            v.dtype() !=
+                                                dtype<bucket<Variable>>;
+                                   });
+      it != vars.end())
+    return it->dtype();
+  // Second pass: Find bucket<Variable> (or other)
+  if (const auto it =
+          std::find_if(vars.begin(), vars.end(),
+                       [](const auto &v) { return variable::is_bins(v); });
+      it != vars.end())
+    return it->dtype();
+  // Not binned data
```

Comment:
Minor: Could maybe refactor this with a sepate function returning the dtype and taking lambda as arg.

Reply:
I don't understand? You mean a wrapper for `std::find_if`? What do we gain from that? I suppose you mean something else?

Reply:
Not a wrapper for `find_if`. I looks like you are extracting the bins dtype from a vector of VariableConstView. Only difference between first attempt and second pass is the lambda used. It probably wouldn't reduce the number of lines by extracting a separate function for this, so leave as is.